### PR TITLE
[WIP] Static default scheduler

### PR DESCRIPTION
**Pull request by davidwdan**
The purpose of this PR is discuss using a default static scheduler.

usage:

``` php
$scheduler = \Rx\Scheduler::getDefault();  //ImmediateScheduler

//change the default
$scheduler = new EventLoopScheduler($loop);
\Rx\Scheduler::setDefault($scheduler);
```

### CODE REVIEW

File: lib/Rx/Observable/TimerObservable.php [hunk 1]
```diff
@@ -25,11 +26,10 @@ public function __construct($dueTime, SchedulerInterface $scheduler = null)
         $this->scheduler = $scheduler;
     }
 
-    public function subscribe(ObserverInterface $observer, SchedulerInterface $scheduler = null)
+    public function subscribe(ObserverInterface $observer)
```
Comment (asm89) [inline lib/Rx/Observable/TimerObservable.php:29]: Why did you remove the scheduler here?

File: lib/Rx/Observable/TimerObservable.php [hunk 2]
```diff
@@ -25,11 +26,10 @@ public function __construct($dueTime, SchedulerInterface $scheduler = null)
...
     {
-        if ($this->scheduler !== null) {
-            $scheduler = $this->scheduler;
-        }
+        $scheduler = $this->scheduler ?: Scheduler::getDefault();
```
Comment (asm89) [inline lib/Rx/Observable/TimerObservable.php:31]: We should probably get rid of `$this->scheduler` here anyway.

**Comment by asm89** [comment]
The idea behind this looks fine to me. Any opinions @mbonneau, @ocramius and @cboden?

**Comment by Ocramius** [comment]
@asm89 this led to incredibly hard-to-debug and coupling in the ZF2 event manager some time ago. Is this just to share the event loop?

**Comment by asm89** [comment]
@Ocramius No, not just to share the event loop, but related I guess. It's mostly a convenience helper so you don't have to pass an explicit scheduler all over the place. That's currently already "supported" but it will then fallback to the `ImmediateScheduler`. That's nice for demos and presentation, but for the real world it doesn't really work well.

**Comment by Ocramius** [comment]
> so you don't have to pass an explicit scheduler all over the place

Makes sense then

File: lib/Rx/Observable/TimerObservable.php (hunk 2 above)
Comment (mbonneau) [inline lib/Rx/Observable/TimerObservable.php:31]: I think it is important to have a `$scheduler` parameter on the constructor of time-based operators. If an unsuspecting user subscribes to an observable that uses timing, they may not know to send the appropriate scheduler.

File: lib/Rx/Subject/ReplaySubject.php [hunk 3]
```diff
@@ -6,6 +6,7 @@
 use Rx\Disposable\CallbackDisposable;
 use Rx\Observer\ScheduledObserver;
 use Rx\ObserverInterface;
+use Rx\Scheduler;
 use Rx\Scheduler\ImmediateScheduler;
```
Comment (mbonneau) [inline lib/Rx/Subject/ReplaySubject.php:10]: Unused use statement.

File: lib/Rx/Observable/TimerObservable.php (hunk 1 above)
Comment (cboden) [inline lib/Rx/Observable/TimerObservable.php:29]: The scheduler should already be decided by the observable already, not on subscription, I would have thought?

**Comment by cboden** [comment]
My initial reaction reflects @Ocramius'.  It's a compromise between ease of use and "the right way". I'd prefer to see a separate handler for default static stuff than the Scheduler handling its own state AND a global static state.

File: lib/Rx/Observable/TimerObservable.php (hunk 1 above)
Comment (mbonneau) [inline lib/Rx/Observable/TimerObservable.php:29]: Right now schedulers "flow" up the chain getting passed from lower operators as they subscribe.

**Comment by davidwdan** [comment]
This is being implemented in the 2.x branch and is superceded by #114
